org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/ObjectComparision.java

```diff
@@ -0,0 +1,654 @@
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.internal.debug.ui;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+import org.eclipse.debug.core.DebugException;
+import org.eclipse.debug.core.model.IVariable;
+import org.eclipse.jdt.debug.core.IJavaArray;
+import org.eclipse.jdt.debug.core.IJavaFieldVariable;
+import org.eclipse.jdt.debug.core.IJavaObject;
+import org.eclipse.jdt.debug.core.IJavaValue;
+import org.eclipse.jdt.debug.core.IJavaVariable;
+import org.eclipse.jdt.internal.debug.core.model.JDIArrayValue;
+import org.eclipse.jdt.internal.debug.core.model.JDINullValue;
+
+public class ObjectComparision {
+
+	public static final String IMMEDIATE_RESULT_KEY = "ImmediateResult"; //$NON-NLS-1$
+	public static final String RESULT_FAILED = "ResultFailed"; //$NON-NLS-1$
+
+	public static final String SELECTED_OBJECT_1 = "Selection1"; //$NON-NLS-1$
+	public static final String SELECTED_OBJECT_2 = "Selection2"; //$NON-NLS-1$
+
+	public static final String KEYSET_1 = "keySet1"; //$NON-NLS-1$
+	public static final String KEYSET_2 = "keySet2"; //$NON-NLS-1$
+
+	public static final String VALUESET_1 = "valueSet1"; //$NON-NLS-1$
+	public static final String VALUESET_2 = "valueSet2"; //$NON-NLS-1$
+
+	/**
+	 * Extracts the actual value of the given object
+	 *
+	 * @param value
+	 *            IJavaValue of the selected object
+	 * @throws DebugException
+	 * @return returns actual value of the object or return combination object's name and id if extraction is not possible
+	 */
+	@SuppressWarnings("nls")
+	private String objectExtraction(IJavaValue value) throws DebugException {
+		if (value.getReferenceTypeName().equals("java.lang.String")) {
+			return value.toString();
+		}
+		if (value.getReferenceTypeName().startsWith("java.lang.StringBu")) {
+			return stringObjectExtraction(value);
+		}
+		for (IVariable v : value.getVariables()) {
+			if (v.getName().equalsIgnoreCase("value")) {
+				return v.getValue().getValueString();
+			}
+		}
+		return value.toString();
+	}
+
+	/**
+	 * Extracts string value of the given string object
+	 *
+	 * @param value
+	 *            IJavaValue of the selected object
+	 * @throws DebugException
+	 * @return returns actual value of the string or return combination object's name and id if extraction is not possible
+	 */
+	@SuppressWarnings("nls")
+	private String stringObjectExtraction(IJavaValue value) throws DebugException {
+		char c;
+		int i;
+		StringBuilder contentBuilder = new StringBuilder();
+		for (IVariable v : value.getVariables()) {
+			if (v.getName().equalsIgnoreCase("value")) {
+				for (IVariable inner : v.getValue().getVariables()) {
+					i = Integer.parseInt(inner.getValue().toString());
+					if (i == 0) {
+						break;
+					}
+					c = (char) i;
```

Comment:
this code needs explanation and tests why its valid. Why is it possible and makes sense to convert String to int to char? Think about non ascii characters is the integer a codepoint. i don't know with only looking at the code.

Reply:
I will add extra comments in it

Reply:
+ Tests for StringBuilder and StringBuffer is there

Reply:
please also use multi char codepoints in that tests. see https://stackoverflow.com/a/9834993/9549750

Reply:
Thanks @jukzi , I will check this and do the changes required